tests/cli.rs

```diff
@@ -0,0 +1,225 @@
+    let mut cmd = Command::new(iroh);
+    cmd.arg("get-ticket").arg(all_in_one);
+    let cmd = if let Some(ref out) = out {
+        cmd.arg("--out").arg(out)
+    } else {
+        &mut cmd
+    };
+
+    // test get stderr output
+    let get_output = cmd.output()?;
+    assert!(get_output.status.success());
+    match_get_stderr(get_output.stderr)?;
+
+    // test output
+    match out {
+        None => {
+            let expect_content = std::fs::read(path)?;
+            assert_eq!(expect_content, get_output.stdout);
+            Ok(())
+        }
+        Some(out) => compare_files(path, out),
+    }
+}
+
+// Wrapping the [`Child`] process here allows us to impl the `Drop` trait ensuring the provide
+// process is killed when it goes out of scope.
+struct ProvideProcess {
+    child: Child,
+}
+
+impl Drop for ProvideProcess {
+    fn drop(&mut self) {
+        self.child.kill().ok();
+        self.child.try_wait().ok();
+    }
+}
+
+fn compare_files(expect_path: impl AsRef<Path>, got_dir_path: impl AsRef<Path>) -> Result<()> {
+    let expect_path = expect_path.as_ref();
+    let got_dir_path = got_dir_path.as_ref();
+    if expect_path.is_dir() {
+        let paths = std::fs::read_dir(expect_path)?;
+        for entry in paths {
+            let entry = entry?;
+            compare_files(entry.path(), got_dir_path)?;
+        }
+    } else {
+        let file_name = expect_path.file_name().unwrap();
+        let expect = std::fs::read(expect_path)?;
+        let got = std::fs::read(got_dir_path.join(file_name))?;
+        assert_eq!(expect, got);
+    }
+
+    Ok(())
+}
+
+// Looks for regex matches on stderr output for the getter.
+//
+// Errors on the first regex mis-match or if the stderr output has fewer lines than expected
+fn match_get_stderr(stderr: Vec<u8>) -> Result<()> {
+    let stderr = std::str::from_utf8(&stderr[..])?;
+    let mut lines = stderr.lines();
+    let res = vec![
+        r"Fetching: [\da-z]{59}",
+        r"\[1/3\] Connecting ...",
+        r"\[2/3\] Requesting ...",
+        r"\[3/3\] Downloading collection...",
+        r"\d* file\(s\) with total transfer size [\d.]* ?[KMGT]?i?B",
+        r"Done in \d* seconds?",
+    ];
+
+    for re in res {
+        matches(
+            lines
+                .next()
+                .context("Unexpected end of 'get' output")?
+                .trim(),
+            re,
+        )?;
+    }
+    Ok(())
+}
+
+/// Looks for regex matches on stderr output for the provider, returning the "all in one ticket"
+/// that can be used to 'get' from another process.
+///
+/// Errors on the first regex mismatch or if the stderr output has fewer lines than expected
+fn match_provide_stderr(stderr: BufReader<ChildStderr>, use_stdin: bool) -> Result<String> {
+    let mut lines = stderr.lines();
+
+    let mut res = vec![
+        r"Reading \S*",
+        r"Collection: [\da-z]{59}",
+        r"",
+        r"PeerID: [_\w\d-]*",
+        r"Auth token: [\w\d]*",
+    ];
```

Comment:
this can be made even more "general"
right now we iterate through each line & ensure the lines appear in order.
We could instead push all those lines to a single string and have each regex expect to exist _somewhere_ in the output, rather than in this specific order.

Reply:
I think having an order is great

Reply:
in order is fine i guess. otherwise it would just make it more complex